cmake/extensions.cmake

```diff
@@ -65,6 +65,27 @@
 #  zephyr_compile_options()     --->  target_compile_options()
 #
 
+# override the target_sources function to inject a unique ID definition for each source file
+function(target_sources target_name)
+  foreach(src ${ARGN})
+    if(NOT src STREQUAL "PRIVATE" AND
+       NOT src STREQUAL "PUBLIC" AND
+       NOT src STREQUAL "INTERFACE"
+    )
+      # convert to an absolute path to be unique
+      cmake_path(ABSOLUTE_PATH src NORMALIZE)
+      # replace special characters with an underscore
+      string(REGEX REPLACE "[\\\/\.]" "_" unique_id ${src})
+      # add the definition
+      set_property(SOURCE ${src}
```

Comment:
as soon as someone somewhere calls `target_sources()` with a generator expression then this will break.